20240129-satellite-managed-encryption.md

```diff
@@ -0,0 +1,171 @@
+* uses a securely-stored, random passphrase stored on the satellite as the project's encryption passphrase
+* restricts all files to be uploaded without path encryption
+
+The main goals are to:
+* provide users with an experience that does not require managing and entering an encryption passphrase in the Satellite UI
+* provide users with an easy way to view and list files in lexographical order
+
+### Approach / Design
+
+#### Key Management Store
+
+We need a secure way to store and look up project encryption passphrases on the Satellite. Someone with normal Satellite DB access should not be able to view this data, and ideally it is encrypted. At the same time, it must be very reliable, as losing data about project encryption passphrases means that data may have been lost. The primary features and concerns we should keep in mind:
+* as simple as possible from an implementation perspective, e.g. key-value lookup for project ID -> encryption passphrase
+* secure - should be encrypted, and highly restricted who/what has access to it
+* reliable - should have the same standards for uptime/data loss as metainfo DB
+
+The KMS will be made up of 3 components:
+* The master key and store - The master key will be used to encrypt and decrypt passphrases. The store will be used to securely store the master key. The store will be [Google's Secret Manager](https://cloud.google.com/security/products/secret-manager), which is a secure store for secrets such as API keys and other sensitive data.
+This secrets manager was chosen because it is just as secure as the alternatives, if not more and yet cheaper and easier to manage. It also is more convenient to use as we already use GCP.
+* The KMS service - This is part of the Satellite that will be responsible for fetching the master key at startup, encrypting and decrypting passphrases, and storing and retrieving same from the passphrase store. The master key will be kept in memory throughout the lifetime of the Satellite.
+The satellite will be configured with credentials needed to access the Secret Manager (the master key store).
+* The Passphrase store - This where the encrypted passphrases will be stored. It will be a new column, `passphrase_enc` on the `projects` table.
+
+#### Satellite Project Updates
+
+A new flag should be added to the `projects` table, like `projects.satellite_managed_encryption`. This should be false for all existing projects. While satellite-managed-encryption (this design) is in-progress, new projects should be created with this flag set to false.
+Another column should be added to the `projects` table to be the passphrase store, like `projects.passphrase_enc`. This should be the encrypted passphrase for satellite-managed-encryption projects only and null for others.
+
+Functionality should then be added to allow creating a project with a satellite managed encryption passphrase:
+* the `projects.satellite_managed_encryption` column is set to `true` for this project
```

Comment:
Do we need this column if we can check that `passphrase_enc` is `null`?

Reply:
I don't think we need a separate boolean flag if we can use nullable `passphrase_enc` as a flag

Reply:
I've removed that